library/core/src/iter/adapters/enumerate.rs

```diff
@@ -264,3 +251,48 @@ where
 
 #[unstable(issue = "none", feature = "inplace_iteration")]
 unsafe impl<I: InPlaceIterable> InPlaceIterable for Enumerate<I> {}
+
+trait EnumerateSpec<Item> {
+    fn fold<Acc, Fold>(self, init: Acc, fold: Fold) -> Acc
+    where
+        Fold: FnMut(Acc, (usize, Item)) -> Acc;
+}
+
+impl<I: Iterator> EnumerateSpec<I::Item> for Enumerate<I> {
+    #[inline]
+    default fn fold<Acc, Fold>(self, init: Acc, fold: Fold) -> Acc
+    where
+        Fold: FnMut(Acc, (usize, I::Item)) -> Acc,
+    {
+        #[inline]
+        fn enumerate<T, Acc>(
+            mut count: usize,
+            mut fold: impl FnMut(Acc, (usize, T)) -> Acc,
+        ) -> impl FnMut(Acc, T) -> Acc {
+            #[rustc_inherit_overflow_checks]
+            move |acc, item| {
+                let acc = fold(acc, (count, item));
+                count += 1;
+                acc
+            }
+        }
+
+        self.iter.fold(init, enumerate(self.count, fold))
+    }
+}
+
+impl<I: Iterator + TrustedRandomAccessNoCoerce> EnumerateSpec<I::Item> for Enumerate<I> {
```

Comment:
I think, it is worth to add link to the issue in comments because it is not clear why this specialization is needed.

Reply:
Good point, then we can remove it when LLVM is fixed.